### PR TITLE
Add class Substitute

### LINKED ISSUE #1383: Create a Substitute class
Implement a `Substitute` class that has `freeVariables`, `refreshVariables`, `mapVariables`, and `traverseVariables` (which basically support capture-avoiding substitution), and also `substitute`?

### PR DESCRIPTION
**Pull request by ttuegel**
Fixes #1383.

This pull request adds a class `Substitute` which handles substitution and variable renaming.

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

File: kore/src/Kore/Substitute.hs [hunk 1]
```diff
@@ -1,159 +1,56 @@
...
-
-{- | Traverse the pattern from the top down and apply substitutions.
-
-The 'freeVariables' annotation is used to avoid traversing subterms that
-contain none of the targeted variables.
-
-The substitution must be normalized, i.e. no target (left-hand side) variable
-may appear in the right-hand side of any substitution, but this is not checked.
--}
-substitute ::
```
Comment (ana-pantilie) [inline kore/src/Kore/Substitute.hs:44]: This implementation was moved to the `TermLike` instance of `Substitute`, right?

**Review by ana-pantilie — APPROVED** [review]
I just have some minor comments and questions.

**Review by andreiburdusa — APPROVED** [review]

File: kore/kore.cabal [hunk 2]
```diff
@@ -321,6 +321,7 @@ library
     Kore.Internal.Predicate
     Kore.Internal.SideCondition
     Kore.Internal.SideCondition.SideCondition
+    Kore.Substitute
```
[the remark this answers is not in the document]
Comment (ttuegel) [reply kore/kore.cabal]: That was an oversight on my part. The module was removed and re-created, and I put it in the wrong spot.

File: kore/src/Kore/Step/AntiLeft.hs [hunk 3]
```diff
@@ -117,6 +109,28 @@ data AntiLeftLhs variable = AntiLeftLhs
     deriving anyclass (SOP.Generic, SOP.HasDatatypeInfo)
     deriving anyclass (Debug, Diff)
 
+instance InternalVariable variable => Substitute (AntiLeftLhs variable) where
```
[the remark this answers is not in the document]
Comment (ttuegel) [reply kore/src/Kore/Step/AntiLeft.hs:112]: Yes, we use it in `RulePattern`.

File: kore/src/Kore/Substitute.hs (hunk 1 above)
Comment (ttuegel) [reply kore/src/Kore/Substitute.hs:44]: Yes, that's right.

**Pull request merged**